postgresql_user: use standard module hashlib instead of passlib

pilou- opened the pull request:
##### SUMMARY
When `passlib` isn't available, an useless `alter role` query is executed.

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
postgresql_user

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
ansible 2.4.0 (devel 8b4f5ba064)
```

pilou- committed:
postgresql_user: fix doc

See: https://www.postgresql.org/docs/current/static/sql-createrole.html

pilou- committed:
Avoid useless queries: use standard module hashlib

When an unchanged MD5-hashed password was used and passlib was
unavailable, an useless 'ALTER USER' query was executed.

Once this useless query avoided, the last 'SELECT' query becomes
useless too.

pilou- committed:
postgresql_user: add tests with hashed password

mscherer commented on lib/ansible/modules/database/postgresql/postgresql_user.py:
```diff
@@ -52,8 +52,8 @@
         C('str[\\"md5\\"] + md5[ password + username ]'), resulting in a total of 35 characters.  An easy way to do this is:
-        C(echo \\"md5`echo -n \\"verysecretpasswordJOE\\" | md5`\\"). Note that if encrypted is set, the stored password will be hashed whether or not
-        it is pre-encrypted.
+        C(echo \\"md5`echo -n \\"verysecretpasswordJOE\\" | md5`\\"). Note that if the presented password string is already in
```
I think "presented password string" is not correct, shouldn't it be "provided password string" ?

pilou- replied:
Indeed, done.

mscherer commented on lib/ansible/modules/database/postgresql/postgresql_user.py:
```diff
@@ -52,8 +52,8 @@
-        C(echo \\"md5`echo -n \\"verysecretpasswordJOE\\" | md5`\\"). Note that if encrypted is set, the stored password will be hashed whether or not
-        it is pre-encrypted.
+        C(echo \\"md5`echo -n \\"verysecretpasswordJOE\\" | md5`\\"). Note that if the presented password string is already in
+        MD5-encrypted format, then it is used as-is, regardless of encrypted parameter.
```
I think that's "MD5-hashed", crypto people tend to complain about it.

pilou- replied:
Done

mscherer commented on test/integration/targets/postgresql/tasks/main.yml:
```diff
@@ -186,87 +186,15 @@
+    db_password1: '{{ item.password }}'
+  with_items:
+    - encrypted: 'yes'
+      password: 'secretù'
```
Can you add a comment about ù not being a typo, and being used for Unicode testing ?

pilou- replied:
Done

mscherer commented on test/integration/targets/postgresql/tasks/test_user.yml:
```diff
@@ -0,0 +1,92 @@
+    <<: *task_parameters
+    postgresql_user: *parameters
+
+  - name: Check that ansible reports they were created
```
I think that should be "it was created"

pilou- replied:
Done (3 occurrences updated).

mscherer commented on test/integration/targets/postgresql/tasks/main.yml:
```diff
@@ -186,87 +186,15 @@
+- include: test_user.yml
+  vars:
+    encrypted: '{{ item.encrypted }}'
+    db_password1: '{{ item.password }}'
```
Since the value do not change in the loop, wouldn't it be clearer to directly put the password here directly  ?

pilou- replied:
Done

mscherer commented on lib/ansible/modules/database/postgresql/postgresql_user.py:
```diff
@@ -292,22 +294,14 @@ def user_alter(cursor, module, user, password, role_attr_flags, encrypted, expir
-                        # Cannot check if passlib is not installed, so assume password is different
-                        pwchanging = True
-            else:
+            if ((password.startswith('md5') and len(password) == 32+3) or encrypted == 'UNENCRYPTED'):
```
I would document why 32+3 (md5 hash + size of 'MD5').

pilou- replied:
Done

mscherer commented on lib/ansible/modules/database/postgresql/postgresql_user.py:
```diff
@@ -292,22 +294,14 @@ def user_alter(cursor, module, user, password, role_attr_flags, encrypted, expir
                     pwchanging = True
 
+            if not pwchanging and encrypted == 'ENCRYPTED':
+                if md5(to_bytes(password) + to_bytes(user)).hexdigest() != current_role_attrs['rolpassword']:
```
So, if the password start by 'md5', this trigger a special case (that's the whole point of the PR). but why do we compare the md5 of the password + user with current_role_attrs['rolpassword'], since  current_role_attrs['rolpassword'] will be prefixed with 'md5' ? (ie, this would always change the password)

pilou- replied:
You're right. I reworked the whole pull-request (the next condition was wrong too) and improved the tests.

mattclay commented:
CI failure due to traceback during integration tests on python 3:

```
2017-05-01 01:10:45 TASK [postgresql : Create a user (password encrypted: yes)] ********************
2017-05-01 01:10:46 An exception occurred during task execution. To see the full traceback, use -vvv. The error was: UnicodeEncodeError: 'ascii' codec can't encode character '\xf9' in position 6: ordinal not in range(128)
2017-05-01 01:10:46 fatal: [testhost]: FAILED! => {"changed": false, "failed": true, "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/ansible__o2_anvd/ansible_module_postgresql_user.py\", line 770, in <module>\n    main()\n  File \"/tmp/ansible__o2_anvd/ansible_module_postgresql_user.py\", line 729, in main\n    changed = user_add(cursor, user, password, role_attr_flags, encrypted, expires)\n  File \"/tmp/ansible__o2_anvd/ansible_module_postgresql_user.py\", line 269, in user_add\n    cursor.execute(query, query_password_data)\n  File \"/usr/lib/python3/dist-packages/psycopg2/extras.py\", line 120, in execute\n    return super(DictCursor, self).execute(query, vars)\nUnicodeEncodeError: 'ascii' codec can't encode character '\\xf9' in position 6: ordinal not in range(128)\n", "module_stdout": "", "msg": "MODULE FAILURE", "rc": 1}
```

abadger commented:
My guess is that the change in the test cases (maybe because the new password is now non-ascii?) is revealing a bug in a different part of hte code.

pilou- commented:
It seems the second shippable pass with `maximum verbosity` don't fail.

Using locally a Xenial LXC container (`python3-psycopg2: Installed: 2.6.1-1build2`) and the command below, the test suite is successful:
```ANSIBLE_REMOTE_USER=root ANSIBLE_ROLES_PATH=test/integration/targets ansible-playbook -i 'testhost,' -e ansible_host=$IP -e 'ansible_python_interpreter=/usr/bin/python3' --tags=test_postgresql,test_postgresql_db,test_postgresql_privs,test_postgresql_user,needs_privileged test/integration/destructive.yml```.

Using docker, I encountered a problem using `test/runner/ansible-test integration --verbose postgresql --docker ubuntu1604py3`:
```
TASK [setup_postgresql_db : start postgresql service] **************************
fatal: [testhost]: FAILED! => {"changed": false, "cmd": "/bin/systemctl", "failed": true, "msg": "Failed to connect to bus: No such file or directory", "rc": 1, "stderr": "Failed to connect to bus: No such file or directory\n", "stderr_lines": ["Failed to connect to bus: No such file or directory"], "stdout": "", "stdout_lines": []}
```
so I added `--docker-privileged` parameter, then I got
```
TASK [postgresql : Create DB] **************************************************
fatal: [testhost]: FAILED! => {"changed": false, "failed": true, "module_stderr": "/bin/sh: 1: /tmp/ansible-test-coverage-wevrv_99/coverage/runner3.5: Permission denied\n", "module_stdout": "", "msg": "MODULE FAILURE", "rc": 126}
```
do you known how to disable coverage ?

mattclay commented:
I'm guessing the maximum verbosity passing is not due to the verbosity, but the fact that the test is being run a second time.

mattclay commented:
Coverage is off by default. The error you're getting indicates the interpreter interception script used by ansible-test is not executable, which is an error I haven't seen before. Is your `/tmp` directory not allowed to have executable files?

Regarding the docker error, what are you using for your docker host (platform and docker version)?

mattclay commented:
If you don't mind running the tests without docker, you could try this on a system with python 3.5:

`ansible-test integration -v postgresql --allow-destructive --python 3.5`

sivel commented:
I can duplicate the exception with the following command from this branch:

```
ansible-test integration -v --docker ubuntu1604py3 --docker-privileged --python 3.5 postgresql
```

When the except is encountered the following variables are in play (with postgresql_user.py modified to display the following):

```
raise Exception((query, query_password_data))\nException: ('CREATE USER \"ansible_db_user1\" WITH UNENCRYPTED PASSWORD %(password)s ', {'password': 'secretù', 'expires': None})
```

This relates to:

```
db_password1: 'secretù' # use UTF-8
```

sivel commented:
Making the following changes has resolved the issue for me:

```diff
@@ -711,7 +701,9 @@ def main():

     try:
         db_connection = psycopg2.connect(**kw)
+        db_connection.set_client_encoding('UTF8')
         cursor = db_connection.cursor(cursor_factory=psycopg2.extras.DictCursor)
+        psycopg2.extensions.register_type(psycopg2.extensions.UNICODE, cursor)
```

Not sure that `register_type` is required.  The default encoding seems to be `SQLASCII` which cannot handle the unicode character in the password.

pilou- commented:
@mattclay  on a system with python 3.5:

```
# . hacking/env-setup
# ansible-test integration -v postgresql --allow-destructive --python 3.5
/usr/bin/env: ‘python’: No such file or directory
# ln -s /usr/bin/python3 /usr/bin/python
[...]
TASK [postgresql : Create DB] *******************************************************************************************************************************
fatal: [testhost]: FAILED! => {"changed": false, "failed": true, "module_stderr": "/bin/sh: 1: /tmp/ansible-test-coverage-mxg5xz70/coverage/runner3.5: Permission denied\n", "module_stdout": "", "msg": "MODULE FAILURE", "rc": 126}
```

In order to debug, I added a `pause` task before `Create DB`:
```
# ls -l /tmp/ansible-test-coverage-mxg5xz70/coverage/runner3.5
-rwx------ 1 root root 4897 May  4 20:52 /tmp/ansible-test-coverage-mxg5xz70/coverage/runner3.5
```
hence `postgres` user isn't allowed to execute `runner3.5` using the following command:
```
/bin/sh -c 'sudo -H -S -n -u postgres /bin/sh -c '"'"'echo BECOME-SUCCESS-osugeaiedjwqhizxtwcprmekszxwawum; /tmp/ansible-test-coverage-mxg5xz70/coverage/runner3.5 /tmp/ansible-tmp-1493931560.751374-64468188001566/postgresql_db.py'"'"' && sleep 0'
```

Here is a patch proposal: #24311.

@sivel it appears there isn't a default locale defined in `ubuntu1604py3`. `/etc/default/locale` is empty, then `LANG` and `LANGUAGE` environment variables are not set and `LC_*`  are set to `POSIX`. Then PostgreSQL [fallbacks](https://github.com/postgres/postgres/blob/master/src/port/chklocale.c#L297) to ASCII.
I propose to define `PGCLIENTENCODING` environment variable in the test suite rather than modifying the module.

abadger commented:
The postgresql module tests create some locales.  Perhaps that needs to be
applied to these integration tests as well (perhaps abstracted into a
postgresql setup role).

pilou- committed:
Creates cluster the same way Debian package does

This task is only executed when the playbook has already been executed
once, for example using 'ansible-test integration' with '--retry-error'
switch when the first run fails.

This modification allows to recreate default databases (postgres,
template0 and template1) using the same encoding that the one used by
the Debian package.

Default encoding is 'SQL_ASCII' when default locale is not set in
/etc/default/locale.

pilou- committed:
PostgreSQL tests: enforce UTF8

By default, client encoding is determined either from the LANG_*/LC_*
environment variables or using encoding of the database.

Containers used in the CI don't define a default locale, then encoding
of default databases was SQL_ASCII.

pilou- commented:
The commit `Creates cluster the same way Debian package does` allows to encounter the error in the second run of the CI. With the second try, the PostgreSQL databases were initialized differently hence the error was avoided.

This error occurs because in the docker container `ubuntu1604py3`, the default locale isn't set. During installation of the PostgreSQL Debian package, the install script unset `LANG*` and `LC_*` environment variables, then it reads `/etc/default/locale`. Because this file is empty, the script set `LANG` to `C.UTF-8`, then `SQL_ASCII` encoding is used while creating default database.

There are several ways to fix this error, for example:
- set `PGCLIENTENCODING` environment variable when connecting to the database
- configure default locale of container `ubuntu1604py3` just before creating default databases, next unset default locale after databases creation.

sivel commented:
I suppose you could just set `environment` on that single task.

pilou- commented:
Because the tests [set](https://github.com/ansible/ansible/blob/6aa32d3e9d82e9b318c1a150e2c391ac33e469cf/test/runner/lib/util.py#L190) `LC_ALL` to `en_US.UTF-8`, it would make sense to configure the test containers accordingly.

Anyway, I have updated this pull-request in order to set `PGCLIENTENCODING` only when required (7 tasks affected).

pilou- commented:
Pull-request updated (rebased on `devel` branch).

pilou- commented:
Failure on `ubuntu1604py3`:
```
2017-05-12 10:53:19 Status: Downloaded newer image for ansible/ansible:ubuntu1604py3
2017-05-12 10:53:19 Creating a compressed tar archive of path: .
2017-05-12 10:53:23 Resulting archive is 5368841 bytes.
2017-05-12 10:53:23 Run command: docker run --detach ansible/ansible:httptester
2017-05-12 10:53:23 Run command: docker run --detach --volume /sys/fs/cgroup:/sys/fs/cgroup:ro --privileged=false --link 75b64ae516d20a2d1c78deba35b75f017082552d880b2b449cb5cf8edb239b4d:ansible.http.tests --link 75b64ae516d20a2d1c78deba35b75f017082552d880b2b449cb5cf8edb239b4d:sni1.ansible.http.tests --link 75b64ae516d20a2d1c78deba35b75f017082552d880b2b449cb5cf8edb239b4d:sni2.ansible.http.tests --link 75b64ae516d20a2d1c78deba35b75f017082552d880b2b449cb5cf8edb239b4d:fail.ansible.http.tests --env HTTPTESTER=1 --link cloudstack-sim ansible/ansible:ubuntu1604py3
2017-05-12 10:53:24 Run command: docker exec -i d9215f1c307fb2414e85e3816246a720ff55e1340904c67937c9212070753d3c dd of=/root/docker.sh bs=65536
2017-05-12 10:53:24 Run command: docker exec d9215f1c307fb2414e85e3816246a720ff55e1340904c67937c9212070753d3c /bin/bash /root/docker.sh
2017-05-12 10:53:24 /root/docker.sh: line 14: cannot create temp file for here-document: No such file or directory
2017-05-12 10:53:24 Run command: docker rm -f 75b64ae516d20a2d1c78deba35b75f017082552d880b2b449cb5cf8edb239b4d
2017-05-12 10:53:24 Run command: docker rm -f d9215f1c307fb2414e85e3816246a720ff55e1340904c67937c9212070753d3c
2017-05-12 10:53:24 ERROR: Command "docker exec d9215f1c307fb2414e85e3816246a720ff55e1340904c67937c9212070753d3c /bin/bash /root/docker.sh" returned exit status 1.
2017-05-12 10:53:24 Run command: docker rm -f cloudstack-sim
2017-05-12 10:53:25 + cleanup
2017-05-12 10:53:25 + grep -q .
2017-05-12 10:53:25 + find test/results/coverage/ -mindepth 1 -name '.*' -prune -o -print -quit
2017-05-12 10:53:25 + rmdir shippable/testresults/
2017-05-12 10:53:25 + cp -a test/results/junit/ shippable/testresults/
2017-05-12 10:53:25 + cp -aT test/results/bot/ shippable/testresults/
```

mattclay commented:
I've restarted CI.

mscherer commented on test/integration/targets/setup_postgresql_db/tasks/main.yml:
```diff
@@ -63,9 +63,9 @@
-  command: /usr/bin/pg_createcluster {{ pg_ver }} main
-  # Sometimes package install creates the db cluster, sometimes this step is needed
-  ignore_errors: True
+  shell: '. /usr/share/postgresql-common/maintscripts-functions && set_system_locale && /usr/bin/pg_createcluster -u postgres {{ pg_ver }} main'
```
Does it work for all debian version ? (especially since the next release happen soon) ?

pilou- replied:
You're right, this way isn't reliable, packaging might evolve: instead of mimic postinst script, reusing it is better.

I removed this task and added another one which purges PostgreSQL packages if they are installed.

pilou- commented:
Failure on `centos7/1`, `ubuntu1404/1`, `ubuntu1604/1` and `ubuntu1604py3/1 ` due to keyserver (`keyserver.ubuntu.com`) unavailability.

Rebased in order to trigger the CI.

mscherer commented:
shipit

pilou- commented:
bot_status

nerzhul commented:
shipit

pilou- commented:
bot_status

pilou- commented:
Rebased

pilou- commented:
shipit

mscherer commented:
shipit

nerzhul commented:
shipit

pilou- commented:
bot_status

matburt commented:
shipit

abadger commented:
This has gotten enough shipits from various people since the last real change.  Merging.

abadger commented:
merged to devel.
